File: app/src/main/java/com/xephorium/crystalnote/ui/creation/CreationPresenter.java
package com.xephorium.crystalnote.ui.creation;

import android.content.Context;
import android.widget.Toast;

import com.xephorium.crystalnote.data.NoteRepository;
import com.xephorium.crystalnote.data.util.NoteUtils;
import com.xephorium.crystalnote.ui.base.BasePresenter;

public class CreationPresenter extends BasePresenter<CreationView> {

    private Context context;
    private NoteRepository noteRepository;

    public CreationPresenter(CreationView view) {
        attachView(view);
        this.context = (Context) view;
        this.noteRepository = new NoteRepository(context);
    }

    public void handleBackClick() {
        if (getView().isNewNotePopulated())
            Toast.makeText(context, "Draft Discarded", Toast.LENGTH_SHORT).show(); // TODO - Implement Warning
        getView().closeCreationActivity();
    }

    public void handleSaveClick(String noteName, String noteContent) {
        if (!checkNoteValidity(noteName, noteContent))
            return;

        noteRepository.writeToNote(noteName, noteContent);
        getView().closeCreationActivity();
    }

    private boolean checkNoteValidity(String name, String content) {
        if (!getView().isNewNotePopulated()) {
            Toast.makeText(context, "No Changes to Save", Toast.LENGTH_SHORT).show();
            return false;
        } else if (!NoteUtils.isValidNoteName(name)) {
            Toast.makeText(context, "Invalid Note Name", Toast.LENGTH_SHORT).show();
            return false;
        } else if (noteRepository.noteExists(name)) {
            Toast.makeText(context, "Note Already Exists", Toast.LENGTH_SHORT).show();
            return false;
        }
        Toast.makeText(context, "Note Saved", Toast.LENGTH_SHORT).show();
        return true;
    }
}

File: app/src/main/java/com/xephorium/crystalnote/ui/creation/CreationPresenter.kt
package com.xephorium.crystalnote.ui.creation

import com.xephorium.crystalnote.data.validation.NoteValidator

class CreationPresenter : CreationContract.Presenter() {


    /*--- Action Handling Methods ---*/

    override fun handleNameTextChange(name: String) {
        this.name = name
    }

    override fun handleContentTextChange(content: String) {
        this.content = content
    }

    override fun handleBackClick() {
        // TODO - Implement Warning
        if (isNoteValid()) view?.showDraftDiscardedMessage()
        view?.navigateBack()
    }

    override fun handleSaveClick() {
        if (!isNoteValid()) {
            view?.showInvalidNoteMessage(getInvalidNoteMessage())
        } else {
            noteRepository.writeToNote(name, content)
            view?.navigateBack()
        }
    }


    /*--- Private Methods ---*/

    private fun isNoteValid(): Boolean = NoteValidator.isValidNoteName(name)
            && NoteValidator.isValidNoteContent(content)
            && !noteRepository.noteExists(name)

    private fun getInvalidNoteMessage() : String {
        return when {
            name.isBlank() && content.isBlank() -> "No Changes to Save"
            !NoteValidator.isValidNoteName(name) -> "Invalid Note Name"
            noteRepository.noteExists(name) -> "Note Already Exists"
            else -> "Note Invalid"
        }
    }
}
